library/std/src/sys/unix/process/process_unix.rs

```diff
@@ -111,6 +104,94 @@ impl Command {
+        // Note that as soon as we're done with the fork there's no need to hold
+        // a lock any more because the parent won't do anything and the child is
+        // in its own process.
+        let _env_lock = unsafe { sys::os::env_lock() };
+
+        // If we fail to create a pidfd for any reason, this will
+        // stay as -1, which indicates an error
+        let mut pidfd: libc::pid_t = -1;
+
+        // On Linux, attempt to use the `clone3` syscall, which
+        // supports more arguments (in particular, the ability to create a pidfd).
+        // If this fails, we will fall through this block to a call to `fork()`
+        cfg_if::cfg_if! {
+            if #[cfg(target_os = "linux")] {
+                static HAS_CLONE3: AtomicBool = AtomicBool::new(true);
+
+                const CLONE_PIDFD: u64 = 0x00001000;
+
+                #[repr(C)]
+                struct clone_args {
+                    flags: u64,
+                    pidfd: u64,
+                    child_tid: u64,
+                    parent_tid: u64,
+                    exit_signal: u64,
+                    stack: u64,
+                    stack_size: u64,
+                    tls: u64,
+                    set_tid: u64,
+                    set_tid_size: u64,
+                    cgroup: u64,
+                }
+
+                syscall! {
+                    fn clone3(cl_args: *mut clone_args, len: libc::size_t) -> libc::c_long
+                }
+
+                if HAS_CLONE3.load(Ordering::Relaxed) {
+                    let mut flags = 0;
+                    if self.make_pidfd {
+                        flags |= CLONE_PIDFD;
+                    }
+
+                    let mut args = clone_args {
+                        flags,
+                        pidfd: &mut pidfd as *mut libc::pid_t as u64,
+                        child_tid: 0,
+                        parent_tid: 0,
+                        exit_signal: libc::SIGCHLD as u64,
+                        stack: 0,
+                        stack_size: 0,
+                        tls: 0,
+                        set_tid: 0,
+                        set_tid_size: 0,
+                        cgroup: 0
+                    };
+
+                    let args_ptr = &mut args as *mut clone_args;
+                    let args_size = crate::mem::size_of::<clone_args>();
+
+                    let res = cvt(unsafe { clone3(args_ptr, args_size) });
+                    match res {
+                        Ok(n) => return Ok((n, pidfd)),
+                        Err(e) => match e.raw_os_error() {
+                            // Multiple threads can race to execute this store,
+                            // but that's fine - that just means that multiple threads
+                            // will have tried and failed to execute the same syscall,
+                            // with no other side effects.
+                            Some(libc::ENOSYS) => HAS_CLONE3.store(false, Ordering::Relaxed),
```

Comment:
Just checking for ENOSYS is sadly not enough here.
The syscall can also be blocked by seccomp (EPERM).

We have also found cases where new syscalls where randomly broken on some configurations (see the EOPNOTSUPP comment [here](https://github.com/rust-lang/rust/blob/4940dd483a8448c0f1ef28d304fad88a9d983c4e/library/std/src/sys/unix/kernel_copy.rs#L580). Though, I think we can ignore this case for now.

Reply:
EPERM and EOPNOTSUPP might depend on the arguments of the clone3 call. It would be wrong to set HAS_CLONE3 to false if one of these errors occurred.

Reply:
In that case the usual approach is to perform the probe separately with arguments that are known to fail with a specific error code, e.g. an invalid combination of flags leading to `EINVAL`. If that returns `EPERM` then we know the syscall is filtered. `CLONE_SIGHAND | CLONE_CLEAR_SIGHAND` could work.

Reply:
Maybe we could just fallback to `fork` on `EPERM` without setting `HAS_CLONE3`.

Reply:
My concern was whether it would be possible for clone3 to succeed after having returned `EPERM` once. In case of `ENOSYS` we can be certain that any future calls will fail as well. For `EPERM`, this conclusion does potentially not hold since a seccomp filter could be defined in terms of arbitrary rules.

Reply:
I don't think we need to account for arbitrary seccomp filters here. seccomp filters *should* produce ENOSYS, not EPERM.